manuals/fsntfsmount.1
.Dd July 27, 2026
.Dt FSNTFSMOUNT 1
.Os
.Sh NAME
.Nm fsntfsmount
.Nd mounts a New Technology File System (NTFS) volume
.Sh SYNOPSIS
.Nm fsntfsmount
.Op Fl o Ar offset
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm fsntfsmount
is a utility to mount a New Technology File System (NTFS) volume
.Pp
.Nm fsntfsmount
is part of the
.Nm libfsntfs
package.
.Nm libfsntfs
is a library to access the New Technology File System (NTFS) format
.Pp
.Ar source
a New Technology File System (NTFS) volume
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl v
verbose output to stderr, while fsntfsmount will remain running in the \
foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fsntfsmount /dev/sda1 /mnt/fsntfs
fshfsmount 20260727
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr fsntfsinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfsntfs/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* pyfsntfs convert lookup path to support additional character range

* libfsntfs and fsntfsmount how to support \\ in file name? since \\ is separator used by the library

* extent: flag WofCompressedData extent as wof_compressed?
* extent: rename compressed flag to lznt1_compressed?

* fsntfsinfo: print volume size
* fsntfsinfo: print extents/data runs of $DATA stream?

* get_file_entry_by_path determine unicode form

* fsntfsmount: cache last 3 recently used directories?

* test if format version 1.2 is supported and updated README

* refactor $I30 index to file entry
** refactor libfsntfs_mft_entry_has_directory_entries_index

* Ntfs compressed stream add uncompressed size check
* add raw (no valid size) data stream support
* check data size of WOF reparse point
* add has WOF compressed data stream function (for dfVFS)

* complete read/write lock support
** refactor volume and file entry to use file system
** file entry
*** libfsntfs_file_entry_get_name_attribute_index
*** libfsntfs_file_entry_get_security_descriptor_size
*** libfsntfs_file_entry_get_security_descriptor
*** continue with libfsntfs_file_entry_get_sub_file_entry_by_index
** MFT metadata file
** specific attribute values functions
** USN journal

* MFT entry
** preserve attribute list data MFT entries?
** libfsntfs_attribute add function to indicate partial attribute (has attribute_list_entry but no mft_attribute)

* bitmap values
** handle sparse data runs more efficiently

* fsntfsinfo complete bodyfile support
** fix libfsntfs_file_entry_get_security_descriptor_size: unable to retrieve security descriptor from index for identifier: 0.
** add support for tsk metadata address calculation? inside MFT attribute?

* fsntfsinfo -F add support for ADS notation

* API changes
** rename libfsntfs_volume_get_number_of_file_entries to libfsntfs_volume_get_number_of_mft_entries

* change libfsntfs_mft_attribute_get_utf[8,16]_name[_size] to return 0 if not set

* add debug compare of long and short names

* refactor index to better handle get entry by identifier
** make index value cache a passthrough for index node cache
** use tree for get entry by filename
** use tree for get entry by security identifier
** remove the need for a local cache in libfsntfs_index_read_sub_nodes, use cache value lock instead

* rename libfsntfs_internal_attribute_get_data to libfsntfs_internal_attribute_get_resident_data
* libfsntfs_attribute add read / seek functions
* change calls to libfsntfs_mft_entry_get_index_by_utf8_name to utf16 for optimization
* change calls to libfsntfs_mft_entry_get_data_attribute_by_utf8_name to utf16 for optimization
* enforce LIBFSNTFS_FILE_ENTRY_FLAGS_MFT_ONLY when trying to read data runs

* libfsntfs_volume_get_root_directory return 0 if MFT entry 5 does not exists

* Check if mft_entry should be cloned before passing to
  libfsntfs_file_entry_initialize in case it is cached out
  otherwise remove it from file_entry

* tests: pass mft_metadata_file different input files
* look into handling directories with high number of entries
* libfsntfs/pyfsntfs attribute
  - expose raw attribute data?
* USN change journal
  - add support for values in $Max
* resident data
  - handle valid data size
* enforce LIBFSNTFS_FILE_ENTRY_FLAGS_MFT_ONLY when trying to read data runs
* fix or remove TODO_CASE_INSENSITIVE

* compressed data
  - handle decompression errors e.g.
    compression tuple offset value out of bounds.
  - have means to map sparse ranges outside compressed blocks
  - handle valid data size
* create fsntfsexport tool to export file entry data
* handle pre 3.1 attribute names?
* check file entry life times
* set file entry name, if get by index and number of filenames is 1

* cached indexes?
* data runs block
  - add a check to see if data run sizes add up
* volume
  - complete libfsntfs_volume_has_bitlocker_drvie_encryption
  - complete libfsntfs_volume_has_volume_shadow_snapshots
* file entry
  - add API get file name by attribute index functions?
  - add API get parent file_reference name by attribute index functions?
  - add API function for owner
  - add API function for security identifier
  - only retrieve file name from the directory entry?
    or make clear in function documentation that first
    non-short filename is the fallback?
  - add support to retrieve name of the root entry if directory entry is missing
* pyfsntfs
  - add attribute type enum
* pyfsntfs file entry
  - expose more MFT entry values through API and pyfsntfs
  - add is empty function
  - add has_reparse_point function
  - add has_symbolic_link function
  - add is_corrupt function?
  - how to handle file-like object functions for non-data file entry
* pyfsntfs file entry/data stream
  - add extents iterator?
* libfsntfs/pyfsntfs file entry
  - expose default data stream
  - add support for short file name
  - fallback to short file name if no long
* libfsntfs_mft_entry_read_element_data set MFT only flag using element flags
* libfsntfs_attribute_read_from_mft handle attribute size out of bounds
* mft metadata file parsing mark file entry as in complete if data is stored
  outside the $MFT
* change MFT entry flags value checks to named definitions
* attribute data parsing
  - add SSIZE_MAX check for attribute->data_size before (size_t) cast

* cached approach of reading attributes ?

* fix issue on win2k image
  - stack overflow infinite loop
  - determine MFT entry index based on offset
    do not rely on value in structure

* MFT entry create get data segment function for internal usage

* how to handle named index roots e.g. $ObjId:$O, $Quota:$O
* fix issue with bounded MFT entry cache
* support MFT entry 0 consisting of multiple $DATA attributes
  - add segment support to ntfs attribute data reference

* file_entry: clone directory entry ?
* what about ADS support ?
  - todo read from ads
* change libfsntfs_attribute_reference_append_attribute into insert and sort
  attribute ranges


fsntfstools/fsntfsmount.c
/*
 * Mounts a New Technology File System (NTFS) volume.
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fsntfstools_getopt.h"
#include "fsntfstools_i18n.h"
#include "fsntfstools_libcerror.h"
#include "fsntfstools_libclocale.h"
#include "fsntfstools_libcnotify.h"
#include "fsntfstools_libfsntfs.h"
#include "fsntfstools_output.h"
#include "fsntfstools_signal.h"
#include "fsntfstools_unused.h"
#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"

mount_handle_t *fsntfsmount_mount_handle = NULL;
int fsntfsmount_abort                    = 0;

/* Signal handler for fsntfsmount
 */
void fsntfsmount_signal_handler(
      fsntfstools_signal_t signal FSNTFSTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fsntfsmount_signal_handler";

	FSNTFSTOOLS_UNREFERENCED_PARAMETER( signal )

	fsntfsmount_abort = 1;

	if( fsntfsmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     fsntfsmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fsntfsmount to mount a New Technology File System (NTFS) volume.";

	fsntfstools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'v', NULL, "verbose output to stderr, while fsntfsmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "volume", "a New Technology File System (NTFS) volume" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libfsntfs_error_t *error                    = NULL;
	system_character_t *option_offset           = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "fsntfsmount";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( fsntfstools_option_t ) );
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
	int result                                  = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations fsntfsmount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                        = { program, NULL };
	struct fuse_args fsntfsmount_fuse_arguments = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args fsntfsmount_fuse_arguments = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *fsntfsmount_fuse_channel  = NULL;
#endif
	struct fuse *fsntfsmount_fuse_handle        = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS fsntfsmount_dokan_operations;
	DOKAN_OPTIONS fsntfsmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fsntfstools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fsntfstools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fsntfstools_output_version_fprint(
	 stdout,
	 program );

	if( fsntfstools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fsntfstools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fsntfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				fsntfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fsntfstools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		fsntfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		fsntfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libfsntfs_notify_set_stream(
	 stderr,
	 NULL );
	libfsntfs_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &fsntfsmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     fsntfsmount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
	if( mount_handle_open(
	     fsntfsmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		fsntfsmount_fuse_arguments.argc = 0;
		fsntfsmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &fsntfsmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fsntfsmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fsntfsmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &fsntfsmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	fsntfsmount_fuse_operations.open       = &mount_fuse_open;
	fsntfsmount_fuse_operations.read       = &mount_fuse_read;
	fsntfsmount_fuse_operations.release    = &mount_fuse_release;
	fsntfsmount_fuse_operations.opendir    = &mount_fuse_opendir;
	fsntfsmount_fuse_operations.readdir    = &mount_fuse_readdir;
	fsntfsmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	fsntfsmount_fuse_operations.getattr    = &mount_fuse_getattr;
	fsntfsmount_fuse_operations.readlink   = &mount_fuse_readlink;
	fsntfsmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	fsntfsmount_fuse_handle = fuse_new(
	                           &fsntfsmount_fuse_arguments,
	                           &fsntfsmount_fuse_operations,
	                           sizeof( struct fuse_operations ),
	                           fsntfsmount_mount_handle );

	if( fsntfsmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          fsntfsmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	fsntfsmount_fuse_channel = fuse_mount(
	                            mount_point,
	                            &fsntfsmount_fuse_arguments );

	if( fsntfsmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	fsntfsmount_fuse_handle = fuse_new(
	                           fsntfsmount_fuse_channel,
	                           &fsntfsmount_fuse_arguments,
	                           &fsntfsmount_fuse_operations,
	                           sizeof( struct fuse_operations ),
	                           fsntfsmount_mount_handle );

	if( fsntfsmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          fsntfsmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 fsntfsmount_fuse_handle );

	fuse_destroy(
	 fsntfsmount_fuse_handle );

	fuse_opt_free_args(
	 &fsntfsmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &fsntfsmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &fsntfsmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	fsntfsmount_dokan_options.Version    = DOKAN_VERSION;
	fsntfsmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	fsntfsmount_dokan_options.SingleThread = TRUE;
#else
	fsntfsmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		fsntfsmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		fsntfsmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	fsntfsmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	fsntfsmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	fsntfsmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	fsntfsmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	fsntfsmount_dokan_operations.CreateDirectory      = NULL;
	fsntfsmount_dokan_operations.Cleanup              = NULL;
	fsntfsmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fsntfsmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fsntfsmount_dokan_operations.WriteFile            = NULL;
	fsntfsmount_dokan_operations.FlushFileBuffers     = NULL;
	fsntfsmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fsntfsmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fsntfsmount_dokan_operations.FindFilesWithPattern = NULL;
	fsntfsmount_dokan_operations.SetFileAttributes    = NULL;
	fsntfsmount_dokan_operations.SetFileTime          = NULL;
	fsntfsmount_dokan_operations.DeleteFile           = NULL;
	fsntfsmount_dokan_operations.DeleteDirectory      = NULL;
	fsntfsmount_dokan_operations.MoveFile             = NULL;
	fsntfsmount_dokan_operations.SetEndOfFile         = NULL;
	fsntfsmount_dokan_operations.SetAllocationSize    = NULL;
	fsntfsmount_dokan_operations.LockFile             = NULL;
	fsntfsmount_dokan_operations.UnlockFile           = NULL;
	fsntfsmount_dokan_operations.GetFileSecurity      = NULL;
	fsntfsmount_dokan_operations.SetFileSecurity      = NULL;
	fsntfsmount_dokan_operations.GetDiskFreeSpace     = NULL;
	fsntfsmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fsntfsmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	fsntfsmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	fsntfsmount_dokan_operations.Cleanup              = NULL;
	fsntfsmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fsntfsmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fsntfsmount_dokan_operations.WriteFile            = NULL;
	fsntfsmount_dokan_operations.FlushFileBuffers     = NULL;
	fsntfsmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fsntfsmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fsntfsmount_dokan_operations.FindFilesWithPattern = NULL;
	fsntfsmount_dokan_operations.SetFileAttributes    = NULL;
	fsntfsmount_dokan_operations.SetFileTime          = NULL;
	fsntfsmount_dokan_operations.DeleteFile           = NULL;
	fsntfsmount_dokan_operations.DeleteDirectory      = NULL;
	fsntfsmount_dokan_operations.MoveFile             = NULL;
	fsntfsmount_dokan_operations.SetEndOfFile         = NULL;
	fsntfsmount_dokan_operations.SetAllocationSize    = NULL;
	fsntfsmount_dokan_operations.LockFile             = NULL;
	fsntfsmount_dokan_operations.UnlockFile           = NULL;
	fsntfsmount_dokan_operations.GetFileSecurity      = NULL;
	fsntfsmount_dokan_operations.SetFileSecurity      = NULL;
	fsntfsmount_dokan_operations.GetDiskFreeSpace     = NULL;
	fsntfsmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fsntfsmount_dokan_operations.Unmounted            = NULL;
	fsntfsmount_dokan_operations.FindStreams          = NULL;
	fsntfsmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &fsntfsmount_dokan_options,
	          &fsntfsmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount New Technology File System (NTFS) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( fsntfsmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 fsntfsmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &fsntfsmount_fuse_arguments );
#endif
	if( fsntfsmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &fsntfsmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

